File: src/main/golang/leetcode/editor/cn/[139]Word Break.go
package lc

//leetcode submit region begin(Prohibit modification and deletion)
func wordBreak(s string, wordDict []string) bool {
	dictset := make(map[string]bool)
	for _, word := range wordDict {
		dictset[word] = true
	}
	n := len(s)
	dp := make([]bool, n+1)
	dp[0] = true
	for i := 1; i <= n; i++ {
		for j := 0; j < i; j++ {
			if dp[j] && dictset[s[j:i]] {
				dp[i] = true
				break
			}
		}
	}
	return dp[n]
}

//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[139]Word Break.py
from typing import List


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        word_dict = {word for word in wordDict}
        n = len(s)
        dp = [False] * (n + 1)
        dp[0] = True
        for i in range(1, n + 1):
            for j in range(i):
                dp[i] |= dp[j] and s[j:i] in word_dict
        return dp[-1]
# leetcode submit region end(Prohibit modification and deletion)
